### PR TITLE
Correct Reason parameter in History-Info header

### LINKED ISSUE #873: Reason Header Field of History-Info Header added for CDIV is malformed.
It appears that the Reason Header Field of a History-Info added for CDIV is malformed. IR-92 references TS 24.604 (4.5.2.6.2.2, (b)) which says the following for the History-Info header: “If the diversion is based on a SIP response from the served user, a Reason header in escaped form shall be included in accordance with IETF RFC 7044 [3].” Looking at 7044 it shows an example of a “embedded” Reason header in section 5 (```History-Info: <sip:UserA@ims.example.com?Reason=SIP%3Bcause%3D302>;index=1.1,…```) and points to 3326 for details. RFC 3326 defines the grammar for the Reason Header Field in section 2 (which lines up with the example in 7044). When Sprout sends a diverted Invite (for busy), the first History-Info header looks like: ```History-Info: <sip:4144910002@vm2.richlab.datcon.co.uk;text=Busy%20Here;cause=486;Reason=SIP>;index=1```. This does not seem to follow the grammar as expected. 

Alex did some digging and found the following: RFC 4244 (the precursor to 7044) contains two different examples for how to encode the cause and text. Section 4.2 uses this form:
```
<sip:UserA@ims.example.com?Reason=SIP%3Bcause%3D302>;index=1;foo=bar
```
However the flows in section 4.5 use this representation, which is the one Sprout is using.
```
<sip:User2@UA2.example.com?Reason=SIP;cause=408;text="RequestTimeout">;index=1.1.1
```
However 7044 only mentions the first form, so I think this is the one we should use. (In fact I think this was probably the intention of the original RFC 4244 – that RFC did not place any special importance on the “cause” or “text” parameters, while RFC 3326 (Reason header) does which makes me think that RFC 4244 intended implementers use the section 4.2 form, and the section 4.5 form was included in error).

### PR DESCRIPTION
**Pull request by chris-elford-metaswitch**
This pull request fixes #873 by correcting the Reason field in the History-Info header. The old behavior was to include each part of the Reason field as a separate parameter in the URI. The fixed behavior is to URL escape the values into a single parameter as per RFC 7044.

I've checked the RFC and confirmed that I agree with the interpretation in the issue. I've run the live tests against my code with no errors.

### CODE REVIEW

File: sprout/pjutils.cpp [hunk 1]
```diff
@@ -1825,25 +1822,35 @@ void PJUtils::update_history_info_reason(pjsip_uri* history_info_uri, pj_pool_t*
...
     {
+      // Need to build up a std::string containing the parameter value
+      // so that it can be manipulated using C++ utility functions.
+      std::stringstream value_builder;
+      std::string reason_value_string;
+      pj_str_t reason_value;
+
+      const pj_str_t* reason_text = pjsip_get_status_text(code);
+
+      value_builder << "SIP;cause=";
```
Comment (rkd-msw) [inline sprout/pjutils.cpp]: Can you chain the `<<`s - `value_builder << "SIP;cause=" << code << ";text=\""...`?

Comment (rkd-msw) [inline sprout/pjutils.cpp]: Or an `snprintf`, or "align your chevrons".

File: sprout/basicproxy.cpp [hunk 2]
```diff
@@ -1733,6 +1733,8 @@ void BasicProxy::UACTsx::cancel_pending_tsx(int st_code)
       if (_tdata->msg->line.req.method.id == PJSIP_INVITE_METHOD)
       {
         TRC_DEBUG("Sending CANCEL request");
+
+        // See issue 1232.
```
Comment (rkd-msw) [inline sprout/basicproxy.cpp]: Not sure this comment adds much value, but I don't mind that much.

File: sprout/bono.cpp [hunk 3]
```diff
@@ -2860,18 +2860,11 @@ void UACTransaction::cancel_pending_tsx(int st_code)
     TRC_DEBUG("Found transaction %s status=%d", name(), _tsx->status_code);
     if (_tsx->status_code < 200)
     {
-      pjsip_tx_data *cancel;
-      pjsip_endpt_create_cancel(stack_data.endpt, _tsx->last_tx, &cancel);
-      if (st_code != 0)
-      {
-        char reason_val_str[96];
```
Comment (rkd-msw) [inline sprout/bono.cpp]: I was a bit nervous about this change but then realised that create_cancel already does this.

File: sprout/pjutils.cpp [hunk 4]
```diff
@@ -1825,25 +1822,35 @@ void PJUtils::update_history_info_reason(pjsip_uri* history_info_uri, pj_pool_t*
     pjsip_sip_uri* history_info_sip_uri = (pjsip_sip_uri*)history_info_uri;
     if (pj_list_empty(&history_info_sip_uri->other_param))
     {
+      // Need to build up a std::string containing the parameter value
+      // so that it can be manipulated using C++ utility functions.
+      std::stringstream value_builder;
+      std::string reason_value_string;
+      pj_str_t reason_value;
+
+      const pj_str_t* reason_text = pjsip_get_status_text(code);
```
Comment (rkd-msw) [inline sprout/pjutils.cpp]: Feels like this is partly common with PJUtils::add_reason - could we refactor the common bit out, to have a single function that takes a status code and returns `SIP;cause=777;text="whatever"`?

**Comment by rkd-msw** [comment]
Looks good - might be worth refactoring to share code with PJUtils::add_reason.

**Comment by rkd-msw** [comment]
(P.S. Convention is to include `[Reviewer: Rob]` in the title as well as assigning)

File: sprout/pjutils.cpp (hunk 1 above)
Comment (chris-elford-metaswitch) [inline sprout/pjutils.cpp]: I've stuck it all on one line. It goes over the 80-limit but it's neater that way.

**Comment by chris-elford-metaswitch** [comment]
Rob, I've followed your suggestion and factored out the reason string building part of the code. I this what you expected?

File: sprout/pjutils.cpp [hunk 5]
```diff
@@ -59,6 +59,11 @@ extern "C" {
 #include "enumservice.h"
 #include "uri_classifier.h"
 
+/// Definitions of static functions.
+namespace PJUtils {
```
Comment (rkd-msw) [inline sprout/pjutils.cpp]: Not worth bothering with the namespace if it's static.

**Comment by rkd-msw** [comment]
One comment, then good to merge.

**Pull request merged**